components/brave_wallet/common/fil_address.cc

```diff
@@ -0,0 +1,163 @@
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+#include "brave/third_party/argon2/src/src/blake2/blake2.h"
+#include "components/base32/base32.h"
+
+namespace brave_wallet {
+
+namespace {
+
+#define SECP256K_PUBLIC_KEY_SIZE 20
+#define SECP256K_ADDRESS_SIZE 41
+#define BLS_ADDRESS_SIZE 86
+#define BLS_PUBLIC_KEY_SIZE 48
+
+std::vector<uint8_t> BlakeHash(const std::vector<uint8_t>& payload,
+                               size_t length) {
+  blake2b_state blakeState;
+  if (blake2b_init(&blakeState, length) != 0) {
+    VLOG(0) << __func__ << ": blake2b_init failed";
+    return std::vector<uint8_t>();
+  }
+  if (blake2b_update(&blakeState, payload.data(), payload.size()) != 0) {
+    VLOG(0) << __func__ << ": blake2b_update failed";
+    return std::vector<uint8_t>();
+  }
+  std::vector<uint8_t> result;
+  result.resize(length);
+  if (blake2b_final(&blakeState, result.data(), length) != 0) {
+    VLOG(0) << __func__ << ": blake2b_final failed";
+    return result;
+  }
+  return result;
+}
+
+bool IsValidNetwork(const std::string& network) {
+  return network == mojom::kFilecoinTestnet ||
+         network == mojom::kFilecoinMainnet;
+}
+
+absl::optional<mojom::FilecoinAddressProtocol> ToProtocol(char input) {
+  if ((input - '0') == static_cast<int>(mojom::FilecoinAddressProtocol::BLS))
+    return mojom::FilecoinAddressProtocol::BLS;
+  if ((input - '0') ==
+      static_cast<int>(mojom::FilecoinAddressProtocol::SECP256K1))
+    return mojom::FilecoinAddressProtocol::SECP256K1;
+  return absl::nullopt;
+}
+}  // namespace
+
+FilAddress::FilAddress(const std::vector<uint8_t>& bytes,
+                       mojom::FilecoinAddressProtocol protocol,
+                       const std::string& network)
+    : protocol_(protocol), network_(network), bytes_(bytes) {
+  DCHECK(IsValidNetwork(network));
+}
+FilAddress::FilAddress() = default;
+FilAddress::FilAddress(const FilAddress& other) = default;
+FilAddress::~FilAddress() {}
+
+bool FilAddress::IsEqual(const FilAddress& other) const {
+  return bytes_.size() == other.bytes_.size() &&
+         std::equal(bytes_.begin(), bytes_.end(), other.bytes_.begin()) &&
+         protocol_ == other.protocol_ && network_ == other.network_;
+}
+
+bool FilAddress::operator==(const FilAddress& other) const {
+  return IsEqual(other);
+}
+
+bool FilAddress::operator!=(const FilAddress& other) const {
+  return !IsEqual(other);
+}
+
+// static
+FilAddress FilAddress::From(const std::string& address) {
+  if (address.size() != BLS_ADDRESS_SIZE &&
+      address.size() != SECP256K_ADDRESS_SIZE)
+    return FilAddress();
+
+  auto protocol = ToProtocol(address[1]);
+  if (!protocol)
+    return FilAddress();
+
+  std::string network{address[0]};
+  if (!IsValidNetwork(network))
+    return FilAddress();
+
+  std::string payload{
+      base32::Base32Decode(base::ToUpperASCII(address.substr(2)))};
+  if (payload.empty())
+    return FilAddress();
+
+  int checksum_size = 4;
```

Comment:
This should be a shared constant, used by `BlakeHash(checksumPayload, 4);` too.

Reply:
done